experimental/traffic-portal/src/app/api/topology.service.ts

```diff
@@ -0,0 +1,196 @@
+	}
+
+	/**
+	 * Creates a new Topology.
+	 *
+	 * @param topology The Topology to create.
+	 */
+	public async createTopology(topology: RequestTopology): Promise<ResponseTopology> {
+		return this.post<ResponseTopology>("topologies", topology).toPromise();
+	}
+
+	/**
+	 * Replaces an existing Topology with the provided new definition of a
+	 * Topology.
+	 *
+	 * @param topology The full new definition of the Topology being updated
+	 */
+	public async updateTopology(topology: ResponseTopology): Promise<ResponseTopology> {
+		return this.put<ResponseTopology>(`topologies?name=${topology.name}`, topology).toPromise();
+	}
+
+	/**
+	 * Generates a material tree from a topology.
+	 *
+	 * @param topology The topology to generate a material tree from.
+	 * @returns a material tree.
+	 */
+	public topologyToTree(topology: ResponseTopology): Array<TopTreeNode> {
+		const treeNodes: Array<TopTreeNode> = [];
+		const topLevel: Array<TopTreeNode> = [];
+		for (const node of topology.nodes) {
+			const name = node.cachegroup;
+			const cachegroup = node.cachegroup;
+			const children: Array<TopTreeNode> = [];
+			const parents: Array<TopTreeNode> = [];
+			treeNodes.push({
+				cachegroup,
+				children,
+				name,
+				parents,
+			});
+		}
+		for (let index = 0; index < topology.nodes.length; index++) {
+			const node = topology.nodes[index];
+			const treeNode = treeNodes[index];
+			if (!(node.parents instanceof Array) || node.parents.length < 1) {
+				topLevel.push(treeNode);
+				continue;
+			}
+			for (const parent of node.parents) {
+				treeNodes[parent].children.push(treeNode);
+				treeNode.parents.push(treeNodes[parent]);
+			}
+		}
+		return topLevel;
+	}
+
+	/**
+	 * Generates a topology from a material tree.
+	 *
+	 * @param name The topology name
+	 * @param description The topology description
+	 * @param treeNodes The data for a material tree
+	 * @returns a material tree.
+	 */
+	public treeToTopology(name: string, description: string, treeNodes: Array<TopTreeNode>): ResponseTopology {
+		const topologyNodeIndicesByCacheGroup: Map<string, number> = new Map();
+		const nodes: Array<ResponseTopologyNode> = new Array<ResponseTopologyNode>();
+		this.treeToTopologyInner(topologyNodeIndicesByCacheGroup, nodes, undefined, treeNodes);
+		const topology: ResponseTopology = {
+			description,
+			lastUpdated: new Date(),
+			name,
+			nodes,
+		};
+		return topology;
+	}
+
+	/**
+	 * Inner recursive function for generating a Topology from a material tree.
+	 *
+	 * @param topologyNodeIndicesByCacheGroup A map of Topology node indices
+	 * using cache group names as the key
+	 * @param topologyNodes The mutable array of Topology nodes
+	 * @param parent The parent, if it exists
+	 * @param treeNodes The data for a material tree
+	 */
+	protected treeToTopologyInner(topologyNodeIndicesByCacheGroup: Map<string, number>,
+		topologyNodes: Array<ResponseTopologyNode>, parent: ResponseTopologyNode | undefined, treeNodes: Array<TopTreeNode>): void {
```

Comment:
you have
```
publicity methodName(arg0,
    arg1, arg2): returnType {
```
which is inconsistent and difficult to read. You _can_ do

```
publicity methodName(arg0,
    arg1,
    arg2): returnType {
```
but IMO that's ugly and I prefer

```
publicity methodName(
    arg0,
    arg1,
    arg2
): returnType {
```
Either one is fine with the linter and keeps it to one argument per line, so do whatever you think is best.

Reply:
Moved to one argument per line